File: app/src/main/java/wangdaye/com/geometricweather/background/polling/PollingUpdateHelper.java
package wangdaye.com.geometricweather.background.polling;

import android.content.Context;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.List;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.basic.models.Location;
import wangdaye.com.geometricweather.common.basic.models.weather.Weather;
import wangdaye.com.geometricweather.common.utils.helpers.AsyncHelper;
import wangdaye.com.geometricweather.common.utils.helpers.BusHelper;
import wangdaye.com.geometricweather.db.DatabaseHelper;
import wangdaye.com.geometricweather.location.LocationHelper;
import wangdaye.com.geometricweather.weather.WeatherHelper;

/**
 * Polling updateRotation helper.
 * */

public class PollingUpdateHelper {

    private final Context mContext;

    private @Nullable AsyncHelper.Controller mIOController;
    private final LocationHelper mLocationHelper;
    private final WeatherHelper mWeatherHelper;

    private List<Location> mLocationList;

    private OnPollingUpdateListener mListener;

    public interface OnPollingUpdateListener {
        void onUpdateCompleted(@NonNull Location location, @Nullable Weather old,
                               boolean succeed, int index, int total);
        void onPollingCompleted(List<Location> locationList);
    }

    public PollingUpdateHelper(Context context,
                               LocationHelper locationHelper,
                               WeatherHelper weatherHelper) {
        mContext = context;

        mIOController = null;
        mLocationHelper = locationHelper;
        mWeatherHelper = weatherHelper;

        mLocationList = null;
    }

    // control.

    public void pollingUpdate() {
        mIOController = AsyncHelper.runOnIO(emitter -> {
            List<Location> list = DatabaseHelper.getInstance(mContext).readLocationList();
            for (Location l : list) {
                l.setWeather(DatabaseHelper.getInstance(mContext).readWeather(l));
            }
            emitter.send(list, true);
        }, (AsyncHelper.Callback<List<Location>>) (locations, done) -> {
            mLocationList = locations;
            requestData(0, false);
        });
    }

    public void cancel() {
        if (mIOController != null) {
            mIOController.cancel();
        }
        mLocationHelper.cancel();
        mWeatherHelper.cancel();
    }

    private void requestData(int position, boolean located) {
        Weather old = mLocationList.get(position).getWeather();
        if (old != null && old.isValid(0.25F)) {
            mLocationList.get(position).setWeather(old);
            new RequestWeatherCallback(old, position, mLocationList.size()).requestWeatherSuccess(
                    mLocationList.get(position));
            return;
        }
        if (mLocationList.get(position).isCurrentPosition() && !located) {
            mLocationHelper.requestLocation(mContext, mLocationList.get(position), true,
                    new RequestLocationCallback(position, mLocationList.size()));
        } else {
            mWeatherHelper.requestWeather(mContext, mLocationList.get(position),
                    new RequestWeatherCallback(old, position, mLocationList.size())
            );
        }
    }

    // interface.

    public void setOnPollingUpdateListener(OnPollingUpdateListener l) {
        mListener = l;
    }

    // on request location listener.

    private class RequestLocationCallback implements LocationHelper.OnRequestLocationListener {

        private final int mIndex;
        private final int mTotal;

        RequestLocationCallback(int index, int total) {
            mIndex = index;
            mTotal = total;
        }

        @Override
        public void requestLocationSuccess(Location requestLocation) {
            mLocationList.set(mIndex, requestLocation);

            if (requestLocation.isUsable()) {
                requestData(mIndex, true);
            } else {
                requestLocationFailed(requestLocation);
                Toast.makeText(
                        mContext,
                        mContext.getString(R.string.feedback_not_yet_location),
                        Toast.LENGTH_SHORT).show();
            }
        }

        @Override
        public void requestLocationFailed(Location requestLocation) {
            if (mLocationList.get(mIndex).isUsable()) {
                requestData(mIndex, true);
            } else {
                new RequestWeatherCallback(null, mIndex, mTotal)
                        .requestWeatherFailed(mLocationList.get(mIndex));
            }
        }
    }

    // on request weather listener.

    private class RequestWeatherCallback implements WeatherHelper.OnRequestWeatherListener {

        private final @Nullable Weather mOld;
        private final int mIndex;
        private final int mTotal;

        RequestWeatherCallback(@Nullable Weather old, int index, int total) {
            mOld = old;
            mIndex = index;
            mTotal = total;
        }

        @Override
        public void requestWeatherSuccess(@NonNull Location requestLocation) {
            mLocationList.set(mIndex, requestLocation);

            Weather weather = requestLocation.getWeather();
            if (weather != null
                    && (mOld == null
                    || weather.getBase().getTimeStamp() != mOld.getBase().getTimeStamp())) {
                BusHelper.postLocationChanged(requestLocation);
                if (mListener != null) {
                    mListener.onUpdateCompleted(requestLocation, mOld, true, mIndex, mTotal);
                }
                checkToRequestNextOrCompleted();
            } else {
                requestWeatherFailed(requestLocation);
            }
        }

        @Override
        public void requestWeatherFailed(@NonNull Location requestLocation) {
            mLocationList.set(mIndex, requestLocation);

            if (mListener != null) {
                mListener.onUpdateCompleted(requestLocation, mOld, false, mIndex, mTotal);
            }
            checkToRequestNextOrCompleted();
        }

        private void checkToRequestNextOrCompleted() {
            if (mIndex + 1 < mTotal) {
                requestData(mIndex + 1, false);
            } else {
                if (mListener != null) {
                    mListener.onPollingCompleted(mLocationList);
                }
            }
        }
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/background/polling/PollingUpdateHelper.kt
package wangdaye.com.geometricweather.background.polling

import android.content.Context
import android.widget.Toast
import wangdaye.com.geometricweather.R
import wangdaye.com.geometricweather.common.basic.models.Location
import wangdaye.com.geometricweather.common.basic.models.weather.Weather
import wangdaye.com.geometricweather.common.bus.EventBus
import wangdaye.com.geometricweather.common.utils.helpers.AsyncHelper
import wangdaye.com.geometricweather.db.DatabaseHelper
import wangdaye.com.geometricweather.location.LocationHelper
import wangdaye.com.geometricweather.weather.WeatherHelper
import wangdaye.com.geometricweather.weather.WeatherHelper.OnRequestWeatherListener

class PollingUpdateHelper(
    private val context: Context,
    private val locationHelper: LocationHelper,
    private val weatherHelper: WeatherHelper
) {
    private var isUpdating = false

    private var ioController: AsyncHelper.Controller? = null
    private var locationList = emptyList<Location>().toMutableList()
    private var listener: OnPollingUpdateListener? = null

    interface OnPollingUpdateListener {

        fun onUpdateCompleted(
            location: Location,
            old: Weather?,
            succeed: Boolean,
            index: Int,
            total: Int
        )
        fun onPollingCompleted(locationList: List<Location>?)
    }

    // control.

    fun pollingUpdate() {
        if (isUpdating) {
            return
        }
        isUpdating = true

        ioController = AsyncHelper.runOnIO({ emitter ->
            val list = DatabaseHelper.getInstance(context).readLocationList().map {
                it.copy(weather = DatabaseHelper.getInstance(context).readWeather(it))
            }
            emitter.send(list, true)
        }, { locations: List<Location>?, _: Boolean ->
            locations?.let {
                locationList = it.toMutableList()
                requestData(0, false)
            }
        })
    }

    fun cancel() {
        isUpdating = false

        ioController?.cancel()
        locationHelper.cancel()
        weatherHelper.cancel()
    }

    private fun requestData(position: Int, located: Boolean) {
        if (locationList[position].weather?.isValid(0.25f) == true) {
            RequestWeatherCallback(position, locationList.size).requestWeatherSuccess(locationList[position])
            return
        }

        if (locationList[position].isCurrentPosition && !located) {
            locationHelper.requestLocation(
                context,
                locationList[position],
                true,
                RequestLocationCallback(position, locationList.size)
            )
            return
        }

        weatherHelper.requestWeather(
            context,
            locationList[position],
            RequestWeatherCallback(position, locationList.size)
        )
    }

    // interface.

    fun setOnPollingUpdateListener(l: OnPollingUpdateListener?) {
        listener = l
    }

    // on request location listener.

    private inner class RequestLocationCallback(
        private val index: Int,
        private val total: Int
    ) : LocationHelper.OnRequestLocationListener {

        override fun requestLocationSuccess(requestLocation: Location) {
            locationList[index] = requestLocation

            if (requestLocation.isUsable) {
                requestData(index, true)
            } else {
                requestLocationFailed(requestLocation)
                Toast.makeText(
                    context,
                    context.getString(R.string.feedback_not_yet_location),
                    Toast.LENGTH_SHORT
                ).show()
            }
        }

        override fun requestLocationFailed(requestLocation: Location) {
            if (locationList[index].isUsable) {
                requestData(index, true)
            } else {
                RequestWeatherCallback(index, total).requestWeatherFailed(locationList[index], false)
            }
        }
    }

    // on request weather listener.

    private inner class RequestWeatherCallback(
        private val index: Int,
        private val total: Int
    ) : OnRequestWeatherListener {

        override fun requestWeatherSuccess(requestLocation: Location) {
            val oldWeather = locationList[index].weather

            if (requestLocation.weather != null
                && (oldWeather == null || requestLocation.weather.base.timeStamp != oldWeather.base.timeStamp)) {
                locationList[index] = requestLocation

                EventBus.instance
                    .with(Location::class.java)
                    .postValue(requestLocation)

                listener?.onUpdateCompleted(requestLocation, oldWeather, true, index, total)

                checkToRequestNextOrCompleted()
            } else {
                requestWeatherFailed(requestLocation, false)
            }
        }

        override fun requestWeatherFailed(requestLocation: Location, apiLimitReached: Boolean) {
            val old = locationList[index].weather
            locationList[index] = requestLocation

            listener?.onUpdateCompleted(requestLocation, old, false, index, total)

            checkToRequestNextOrCompleted()
        }

        private fun checkToRequestNextOrCompleted() {
            if (!isUpdating) {
                return
            }

            val next = index + 1
            if (next < total) {
                requestData(next, false)
                return
            }

            listener?.onPollingCompleted(locationList)
            isUpdating = false
        }
    }
}